java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthVector.java

```diff
@@ -1287,50 +1287,50 @@ public static ArrowBuf set(ArrowBuf buffer, BufferAllocator allocator,
    * @param thisIndex position to copy to in this vector
    * @param from source vector
    */
-  public void copyFrom(int fromIndex, int thisIndex, BaseVariableWidthVector from) {
+  public void copyFrom(int fromIndex, int thisIndex, ValueVector from) {
     if (from.isNull(fromIndex)) {
       fillHoles(thisIndex);
       BitVectorHelper.setValidityBit(this.validityBuffer, thisIndex, 0);
       final int copyStart = offsetBuffer.getInt(thisIndex * OFFSET_WIDTH);
       offsetBuffer.setInt((thisIndex + 1) * OFFSET_WIDTH, copyStart);
     } else {
-      final int start = from.offsetBuffer.getInt(fromIndex * OFFSET_WIDTH);
-      final int end = from.offsetBuffer.getInt((fromIndex + 1) * OFFSET_WIDTH);
+      final int start = from.getOffsetBuffer().getInt(fromIndex * OFFSET_WIDTH);
+      final int end = from.getOffsetBuffer().getInt((fromIndex + 1) * OFFSET_WIDTH);
       final int length = end - start;
       fillHoles(thisIndex);
       BitVectorHelper.setValidityBit(this.validityBuffer, thisIndex, 1);
       final int copyStart = offsetBuffer.getInt(thisIndex * OFFSET_WIDTH);
-      from.valueBuffer.getBytes(start, this.valueBuffer, copyStart, length);
+      from.getDataBuffer().getBytes(start, this.valueBuffer, copyStart, length);
       offsetBuffer.setInt((thisIndex + 1) * OFFSET_WIDTH, copyStart + length);
     }
     lastSet = thisIndex;
   }
 
   /**
-   * Same as {@link #copyFrom(int, int, BaseVariableWidthVector)} except that
+   * Same as {@link #copyFrom(int, int, ValueVector)} except that
    * it handles the case when the capacity of the vector needs to be expanded
    * before copy.
    *
    * @param fromIndex position to copy from in source vector
    * @param thisIndex position to copy to in this vector
    * @param from source vector
    */
-  public void copyFromSafe(int fromIndex, int thisIndex, BaseVariableWidthVector from) {
+  public void copyFromSafe(int fromIndex, int thisIndex, ValueVector from) {
```

Comment:
things can go bad if the caller passes a fixed-width vector in the 'from'.

Reply:
Good question.
Things will go wrong is the source and target are of different vector types. Even if both are  fixed-width vectors, things can go wrong (e.g. from a IntVector to a Float4Vector). 

One solution is to check types each time, however, this may have performance penalty, as copyFrom is a frequently called operation. So current assumption is that the caller should make sure the source and target are of the same type.

Reply:
can we add something cheap, like checking the minorType ?

Reply:
Good suggestion. Thank you @pravindra